staging/src/k8s.io/apiserver/pkg/apis/audit/types.go

```diff
@@ -39,6 +47,22 @@ const (
 	LevelRequestResponse Level = "RequestResponse"
 )
 
+// Stage defines the stages in request handling that audit events may be generated.
+type Stage string
+
+// Valid audit stages.
+const (
+	// The stage for events generated as soon as the audit handler receives the request, and before it
+	// is delegated down the handler chain.
+	StageRequestReceived = "RequestReceived"
+	// The stage for events generated once the response headers are sent, but before the response body
+	// is sent. This stage is only generated for long-running requests (e.g. watch).
+	StageResponseStarted = "ResponseStarted"
+	// The stage for events generated once the response body has been completed, and no more bytes
+	// will be sent.
+	StageResponseComplete = "ResponseComplete"
```

Comment:
- RequestReceived => meta data
- ResponseStarted => header is complete, http status code is filled in, RequestObject is written
- ResponseComplete => ResponseObject is written

Makes sense.

We will create these events:

- short-running reading => ResponseComplete
- short-running writing => RequestReceived + ResponseComplete
- long-running reading => ResponseStarted + ResponseComplete
- long-running writing (e.g. proxy, exec, ...) => all 3
- if we don't know: all three (?)

Does this make sense? Do we need RequestResponseStarted also for short-running-reading?